File: lib/to_mdx/parsers/open_document/style/text_properties.rb
# frozen_string_literal: true

require_relative "../base"

module ToMdx
  module Parsers
    module Style
      class TextProperties < Base
        def initialize(element = nil, context = {})
          super(element, context)
        end

        def to_css
          return {} unless @element&.attributes

          css_properties = {}

          @element.attributes.each do |name, value|
            case name
            when "fo:background-color"
              css_properties["background-color"] = value
            when "fo:color"
              css_properties["color"] = value
            when "fo:font-size"
              value = "#{value.to_f.round}pt" if value.end_with?("pt")
              css_properties["font-size"] = value # TODO based on the default style, adjust this (default = 1rem)
            when "fo:font-weight"
              css_properties["font-weight"] = value unless value == "normal"
            when "fo:font-style"
              css_properties["font-style"] = value unless value == "normal"
            when "style:text-line-through-type"
              css_properties["text-decoration"] = case value
              when "single", "double"
                "line-through"
              else
                nil
              end
            when "style:text-line-through-style"
              css_properties["text-decoration-style"] = case value
              when "solid"
                "solid"
              when "dotted", "dot-dash", "dot-dot-dash"
                "dotted"
              when "dashed", "long-dash"
                "dashed"
              when "wave"
                "wavy"
              else
                nil
              end
            when "style:text-line-through-width"
              css_properties["text-decoration-thickness"] = value
            when "style:font-name"
              css_properties["font-family"] = case value
              when /Mono/
                "'Courier New', Courier, monospace"
              when /Serif/
                "Georgia, 'Times New Roman', Times, serif"
              else
                nil # Assume sans / default
              end

              css_properties["font-weight"] = case value
              when "Bold"
                "bold"
              when "Light"
                "light"
              else
                nil
              end

              css_properties["font-style"] = "italic" if value.include?("Italic")
            when "fo:letter-spacing"
              css_properties["letter-spacing"] = value unless value == "normal"
            end
          end

          css_properties
        end

        register_for "style:text-properties"
      end
    end
  end
end

File: src/parsers/open-document/style/text_properties.ts
import { BaseElement } from "../../base_element.ts";

export class TextProperties extends BaseElement {
  toMdx(): string {
    return "";
  }

  toString(): string {
    return "";
  }

  toCss(): Record<string, string> {
    const cssProperties: Record<string, string> = {};

    Array.from(this.element.attributes).forEach((attr) => {
      const { name, value } = attr;
      switch (name) {
        case "fo:background-color":
          cssProperties["background-color"] = value;
          break;
        case "fo:color":
          cssProperties.color = value;
          break;
        case "fo:font-size":
          cssProperties["font-size"] = value;
          break;
        case "fo:font-weight":
          if (value !== "normal") {
            cssProperties["font-weight"] = value;
          }
          break;
        case "fo:font-style":
          if (value !== "normal") {
            cssProperties["font-style"] = value;
          }
          break;
        case "style:text-line-through-type":
          if (value === "single" || value === "double") {
            cssProperties["text-decoration"] = "line-through";
          }
          break;
        case "style:text-line-through-style":
          switch (value) {
            case "solid":
              cssProperties["text-decoration-style"] = "solid";
              break;
            case "dotted":
            case "dot-dash":
            case "dot-dot-dash":
              cssProperties["text-decoration-style"] = "dotted";
              break;
            case "dashed":
            case "long-dash":
              cssProperties["text-decoration-style"] = "dashed";
              break;
            case "wave":
              cssProperties["text-decoration-style"] = "wavy";
              break;
          }
          break;
        case "style:text-line-through-width":
          cssProperties["text-decoration-thickness"] = value;
          break;
        case "style:font-name":
          if (value.includes("Mono")) {
            cssProperties["font-family"] = "'Courier New', Courier, monospace";
          } else if (value.includes("Serif")) {
            cssProperties["font-family"] = "Georgia, 'Times New Roman', Times, serif";
          }

          if (value === "Bold") {
            cssProperties["font-weight"] = "bold";
          } else if (value === "Light") {
            cssProperties["font-weight"] = "light";
          }

          if (value.includes("Italic")) {
            cssProperties["font-style"] = "italic";
          }
          break;
        case "fo:letter-spacing":
          if (value !== "normal") {
            cssProperties["letter-spacing"] = value;
          }
          break;
      }
    });

    return cssProperties;
  }

  static {
    BaseElement.registerFor.call(this, "style:text-properties");
  }
}
